schema.go

```diff
@@ -445,16 +479,35 @@ func (s *PrimitiveSchema) MarshalJSON() ([]byte, error) {
 	return buf.Bytes(), nil
 }
 
+// Temporary HACK to allow testing schema resolution logic...
+// a better solution would be to extend decoder cache key.
+type primitiveSchemaFingerprint struct {
+	s *PrimitiveSchema
+}
+
+func (sfp *primitiveSchemaFingerprint) String() string {
+	if sfp.s.actual == "" {
+		return sfp.s.String()
+	}
+	return sfp.s.String() + ":" + string(sfp.s.actual)
+}
+
 // Fingerprint returns the SHA256 fingerprint of the schema.
 func (s *PrimitiveSchema) Fingerprint() [32]byte {
-	return s.fingerprinter.Fingerprint(s)
+	return s.fingerprinter.Fingerprint(&primitiveSchemaFingerprint{s: s})
```

Comment:
Decoder cache key logic may require changes to handle primitives with promotion and records with different default values (but same canonical form).

Reply:
As an alternative, I defined a special fingerprint `cacheFingerprint` used only in the context of primitive and record decoders cache.

`cacheFingerprint` is different from the canonical one only if:

- The record has fields with defaults
- The primitive contains an actual type (for promotion)